p2p/peer_behaviour.go

```diff
@@ -0,0 +1,140 @@
+package p2p
+
+import (
+	"errors"
+	"sync"
+)
+
+// ErrorPeerBehaviour are types of observable erroneous peer behaviours.
+type ErrorPeerBehaviour int
+
+const (
+	ErrorPeerBehaviourUnknown = iota
+	ErrorPeerBehaviourBadMessage
+	ErrorPeerBehaviourMessageOutofOrder
+)
+
+// GoodPeerBehaviour are types of observable positive peer behaviours.
+type GoodPeerBehaviour int
+
+const (
+	GoodPeerBehaviourVote = iota + 100
+	GoodPeerBehaviourBlockPart
+)
+
+// PeerBehaviour provides an interface for reactors to signal the behaviour
+// of peers synchronously to other components.
+type PeerBehaviour interface {
+	Behaved(peerID ID, reason GoodPeerBehaviour) error
```

Comment:
I'd go for something stronger than behaved. What we currently have is more like `BehavedForALongTime`, but maybe there's a better word for that.

Reply:
Is it important to encode the temporal part? Sounds very narrow to me.

Reply:
`Behaved` is the normal behaviour. Many peers are behaving even if we're not calling this method. We only call this method once a peer has behaved for a sufficiently long period of time. It's actually stronger than behave ... currently it's a peer we've been receiving valuable unique information from. I'm not sure what a better name would be, `behave` is probably sufficient for now, but we can almost certainly do better :)

Reply:
Also the interface is called `PeerBehaviour` so `Behaved` doesn't feel informative. How did it behave? Maybe this would be better with just `Good` and `Bad` methods?

Reply:
I agree with of issues raised here around naming.  go
[guidelines](https://golang.org/doc/effective_go.html#interface-names)
suggest that interfaces should have `-er` post fix indicating what they
do. If we are reporting peer behaviour the interface should probably be
`PeerReporter`. In which case the only thing the PeerReporter needs
to do is ...Report.  This frees the caller from distinguishing
between good and bad behaviour. Here is one path we could take:

```go
type PeerBehaviour int
const (
	BadMessage = iota
	MessageOutOfOrder
	Vote
	BlockPart
)

type PeerReporter interface {
	Report(peerID ID, reason PeerBehaviour) error
}

type SwitchPeerReporter struct {
	sw *Switch
}

func (spb *SwitchPeerReporter) Report(peerID ID, behaviour PeerBehaviour) error {
	peer := spb.sw.Peers().Get(peerID)
	if peer == nil {
		return errors.New("Peer not found")
	}
    switch behaviour {
    case Vote:
		spb.sw.MarkPeerAsGood(peer, behaviour)))
    case BlockPart:
		spb.sw.MarkPeerAsGood(peer, behaviour))
    case BadMessage:
		spb.sw.StopPeerForError(peer, BlockPart)
    case MessageOutOfOrder:
		spb.sw.StopPeerForError(peer, BlockPart)
    default:
		return errors.New("Unknown Behaviour")
    }

	return nil
}
```

This approach still produces a central set of peer behaviours. I tend to agree with @xla that there is value in having central authority of stopping peers.

Reply:
Maybe `PeerBehaviourReporter`? (it's a bit long but tells what it reports about too). And I wanted to throw in yet another verb/iface name: track / Tracker. As the thing keeps track of peer behaviours 🤔 

I like what @brapse and @xla proposed above. You should probably collapse the cases in the actual code.

Reply:
From the Reactors perspective, I can imagine them taking on responsibility for reporting peers when they witness some behaviour. `Tracking` does have a nice ring to it but to me  it implies some process is being triggered which might not be the case. 
`PeerBehaviourReporter` also sounds nice, and would help distinguish between other reporters. (Metrics perhaps?).

Reply:
Other possibilites: PeerMarker, PeerEvaluator, PeerRater, PeerAppraiser, etc.